### ISSUE TITLE
[Feature Request]: Support neovim

### ISSUE THREAD

**Issue by romgrk** [puremourning/vimspector#29]
I'm ready to send a PR if you're fine with this.

**Comment by puremourning** [comment]
Thanks for that!! Happy to take a look. If it’s complex and likely to break easily, then I might take a view.

**Comment by romgrk** [comment]
Na, won't be complex, I like simple things.

Main issues I see:
 - neovim doesn't support python's `vim.bindeval`
 - neovim/vim job & channel API is different

**Comment by puremourning** [comment]
Does neovim support prompt buffers?

**Comment by romgrk** [comment]
Prompt buffers?

**Comment by puremourning** [comment]
Yes, Vimspector uses them for the watches window and the console window

Here's the vim help on them: https://vimhelp.org/channel.txt.html#prompt-buffer

Basically, you enter insert mode and can type one line, which is then passed to a callback. handy for REPLs.

if you want to pop over to the Gitter room, happy to discuss :)

**Comment by jpfeiffer16** [comment]
For my reference as much as anybody else's, here is the pr for some vim patches that includes prompt buffers.
https://github.com/neovim/neovim/pull/10433

**Comment by puremourning** [comment]
Nice 👍

**Comment by romgrk** [comment]
Tbh, prompt buffers are the easiest part to emulate. Window-toolbars are the real issue :/ But it's nice to not have to emulate them :)

**Comment by puremourning** [comment]
I think we can live without WinBars in NeoVim. People can just use the mappings. Would be satisfied with a PR that didn’t enable them if something like `has( ‘winbar’ )` was false.

**Comment by jpfeiffer16** [comment]
Cool. So far, I haven't found a patch for that submitted against neovim, but maybe we can start with this once the pr for prompt-buffers is merged.

**Comment by romgrk** [comment]
Sounds good to me, I'll continue the PR when prompt-buffers are merged.

**Comment by puremourning** [comment]
:+1:

**Comment by stfl** [comment]
I looks like prompt-buffers are scheduled for nvim 0.5 so that might take awhile as 0.4 just came out.. :(

**Comment by jpfeiffer16** [comment]
I just saw that as well.

**Comment by amustafa** [comment]
Looks like the nightly build is tagged as `0.5`

**Comment by adelarsq** [comment]
There is no plans I think https://github.com/neovim/neovim/issues/8805

**Comment by puremourning** [comment]
https://github.com/puremourning/vimspector/pull/76

**Comment by vn-ki** [comment]
How much is enough people @puremourning?

**Comment by adelarsq** [comment]
@vn-ki some features on neovim side where dropped. So isn't a easy task to support this plugin there. I did some search on this to help, but there are so many things to do that I can't help right now.

**Comment by LinArcX** [comment]
Maybe a fresh project with first class support for neovim?

**Comment by adelarsq** [comment]
There are some (like [vim-vebugger](https://github.com/idanarye/vim-vebugger) and [nvim-gdb](https://github.com/sakhnik/nvim-gdb)), but vimspector is more complete right now.

**Comment by LinArcX** [comment]
No, they don't implement DAP.
I mean fresh new project that implement dap protocol like vimspector. 
I don't have technical info about implementing such a thing, but I can learn and eager to have it on neovim.

**Comment by adelarsq** [comment]
Related: https://github.com/idanarye/vim-vebugger/issues/66

**Comment by puremourning** [comment]
@LinArcX  i wouldn't recommend it - just make vimspector work in neovim ? That seems like the most appropriate thing to do. If you're willing to do the work to make a totally new plugin, why not just support vimspector in neovim ?

**Comment by puremourning** [comment]
@vn-ki to be honest enough people have asked now. there's clearly a real demand. I'll have to evaluate what's required and if i can get solid CI working with neovim. 

I was going to work on windows support next, but i think there's more of a demand for neovim. We can just make a gimped version that doesn't use prompt buffers, and find some workaround for the `bindeval` thing.

**Comment by doronbehar** [comment]
@puremourning Do you think that Creating a `:VimspectorEval` and `:VimspectorWatch` commands as I've suggested (but failed to deliver) in https://github.com/puremourning/vimspector/issues/52#issuecomment-537783029 might be a good start towards "gimping" vimspector for Neovim?

**Comment by puremourning** [comment]
@doronbehar yeah that's what i was thinking.

**Comment by puremourning** [comment]
There is no compatibility between Vim and neovim APIs and feature sets, and the neovim maintainers have said that there's little desire to make them compatible. That includes:

* all job APIs are different (someone wrote async.vim layer, but it's only part of the story)
* all terminal APIs are different
* popup and balloon APIS are different
* channel APIs differ (I _think_, though channels are possibly older than the fork)
* `if_python` (vim python interface) is not 100% compatible
* prompt buffers are missing (there's a PR for it but it's stalled)
* WinBar is missing
* it's marginally harder to make neovim work with python plugins due to its out of process model

Possibly others not yet discovered. So while i would like to do this, it's not going to be quick, and will require a _lot_ of effort and testing. It's going to be hard to prioritise this amount of work and ongoing maintenance.

**Comment by puremourning** [comment]
But never fear, i have a working prototype: https://github.com/puremourning/vimspector/pull/77

Feedback welcome. It's missing bits, but it sorta works.

**Comment by hariamoor** [comment]
@puremourning Bit late to the party, but this might lighten the load on you.

I've seen lots of plugins for Neovim that only work with the master branch (not yet a stable release), but people still heavily use them and give feedback. If things like prompt buffers aren't in a stable release yet, then it shouldn't be a big deal to not support a stable release until they are.

FWIW, many Neovim users compile the master branch from source themselves (like me). The project has diverged from Vim proper and lacks the manpower to prioritize both their own roadmap and Bram's, so users tend to be happy to content themselves with unstable integrations. If you went this route, it wouldn't be a problem.

**Comment by doronbehar** [comment]
> I've seen lots of plugins for Neovim that only work with the master branch (not yet a stable release), but people still heavily use them and give feedback.

I'm not familiar with any such plugins, it'd be nice to hear some examples :)

>  If things like prompt buffers aren't in a stable release yet, then it shouldn't be a big deal to not support a stable release until they are.

Prompt buffers are not only not in stable - they are not even merged to master, see https://github.com/neovim/neovim/pull/10433. More over, even trying to merge this PR locally and then building Neovim with it is not trivial at all,  - the PR needs to be rebased. I've tried doing that a while ago and I've experienced crashes unrelated to this feature.

**Comment by puremourning** [comment]
Have you tried the neovim vimspector branch? It’s feature complete now. No prompt buffers and no Winbar. I added ex commands to replace them.

**Comment by puremourning** [comment]
Oh and it works with the latest release as that’s all I have to test with.

**Comment by hariamoor** [comment]
> I'm not familiar with any such plugins, it'd be nice to hear some examples
@doronbehar The canonical example is [nvim-lsp](https://github.com/neovim/nvim-lsp), which is highly relevant because it's the LSP client that will soon ship with Neovim. I say it's highly relevant because DAP is like the LSP of debuggers.

**Comment by hariamoor** [comment]
I've started an issue in Neovim to get a DAP implementation going to ship with the editor itself, see [neovim/neovim#11732](https://github.com/neovim/neovim/issues/11732).

I don't use Vim, and I'd ideally not like to depend on a non-first-class language for this functionality like Vimspector does on Python. I'm pursuing a Lua implementation of the DAP to be maintained by the Neovim team.

**Comment by puremourning** [comment]
Let’s keep this on topic.

**Comment by puremourning** [comment]
> users tend to be happy to content themselves with unstable integrations.

I doubt that any user is ‘happy’ with an unstable anything. Perhaps ‘tolerant’ is more likely :)

In practice I’m going to have a quick look at what it will take to port the tests to run on neovim, but most likely I will merge neovim support as a ‘best efforts’ feature. So that sort of aligns with what you’re saying, and the experimental nature of this project

**Comment by ovidius72** [comment]
Thanks for the effort. Can you please clarify what is missing for neovim users ?
I see no balloons/tooltip and no WinBar on neovim. Are those the only missing things ? 
Is there any other way to quickly inspect a variable value (without adding it to the watches) in the code window when in pause ? Thanks.

**Comment by puremourning** [comment]
did you see this ? https://github.com/puremourning/vimspector#neovim-differences

I think balloons could be made to work. IIRC vimspector checks for has( 'balloonevalterm' ). it's possible that check should be skipped for neovim.

**Comment by puremourning** [comment]
@ovidius72 I updated the above section to be more explicit about neovim's deficiencies vs Vim-proper.

**Comment by ovidius72** [comment]
Thanks @puremourning 
I saw that in the documentation. I was not sure whether they are missing in neovim or could be used with the `:Vimspector*` command. It would be nice to have balloons replace in neovim with the preview window, and have the change to bind some keys to show it up. 
However, that's really a great plugin.

**Comment by puremourning** [comment]
I'm not going to write yet-more-code just to monkey patch yet more missing features in Neovim. But you can probably make a mapping that does something like `:VimspectorEval <cword>` which would show the value under the cursor in the Console window.

**Comment by puremourning** [comment]
Actually, this works better:

```
:nmap <leader>ve :<c-u>call vimspector#Evaluate( expand( '<cexpr>' ) )<CR>
```

I could add a `<plug>` mapping for that, e.g.:

```
nnoremap <Plug>VimspectorEvalWord
      \ :<c-u>call vimspector#Evaluate( expand( '<cexpr>' ) )<CR>
```

PR welcome for that

**Comment by oblitum** [comment]
> It would be nice to have balloons replace in neovim with the preview window

I'd `s/preview window/floating window/`. There's so many plugins leveraging both vim and neovim floating windows lately, imo balloons have turned into a deprecated artifact.

**Comment by ovidius72** [comment]
@oblitum I just wanted to say that that to emulate the balloons thing in neovim, the preview window could be used. I missed a `d`, it should have been replace**d** :-)

**Comment by puremourning** [comment]
@oblitum  but in neovim I have to write a load of code to achieve that feature. In vim it's like trivial. Making vim use a popup as a balloon? Trivial. There is no all-event mouse tracking in neovim, you end up with having to write a ton of code to manage the popup (whose neovim API is really complex compared to the vim one) and position it, hide it, show it, blah blah blah. I'm just not willing to invest that effort for what should be simple, and is in vim.

**Comment by oblitum** [comment]
@puremourning OK, thing is I've seen so many coc extensions relying on floating windows for widely different purposes, from displaying LSP actions list to alternative spelling display, etc, some of which I use, I thought it wouldn't be that hard. Another thing is that balloons seems tied to mouse? I'm not sure whether it's a mouse feature only, but in that case I'd really not miss it at all as I don't use mouse for nothing when editing/debugging.

**Comment by eyalk5** [comment]
https://github.com/neovim/neovim/issues/2544#issuecomment-528058965

No intention to support popup. Need help in implementing this feature? (hope I have the time)

**Comment by puremourning** [comment]
Well, they haven't made popups easy to use, and they haven't implement all event mouse tracking, so balloons can't be implemented using them.

I would review a PR that does something useful with popups that works in Vim and neovim. But I will not maintain a separate implementation/faeature set just for neovim's lack of certain features.
